### PR TITLE
Candidate and contests tests

### PR DESCRIPTION
**Pull request by gianelgado12**
Add tests for Candidate and Contests data files along with fixing problem with toJsonString methods returning escaped characters.

### CODE REVIEW

File: src/main/java/com/google/sps/servlets/ContestsServlet.java [hunk 1]
```diff
@@ -62,30 +64,30 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
             .map(key -> ServletUtils.getFromDatastore(datastore, key))
-            .map(entity -> Referendum.fromEntity(entity).toJsonString())
+            .map(entity -> JsonParser.parseString(Referendum.fromEntity(entity).toJsonString()))
             .collect(ImmutableList.toImmutableList());
 
     // The toJson function reserializes the already serialized JSON strings, so remove the extra
     // escaped characters.
     Gson gson = new Gson();
-    String contestJson =
-        gson.toJson(contestJsonList).replace("\\", "").replace("\"{", "{").replace("}\"", "}");
```
Comment (anooshrees) [inline src/main/java/com/google/sps/servlets/ContestsServlet.java:86]: I think having escaped quotes within the JSON object is still important in order to maintain cases like Edmund "Jerry" Brown on the site. I used replaceAll("\"","\\\\\"") when reading the Candidate name from the JSON object, but I'll try this as well.

File: .gitignore [hunk 2]
```diff
@@ -10,3 +10,6 @@ config.js
 # IntelliJ specific project files
 .idea/
 portfolio.iml
+
+# vim swap files
+**/*.swp
```
Comment (rogerlaucorp) [inline .gitignore:15]: I prefer small, independent fixes like this to go into separate PRs, which can even skip review from me/Julian. It's fine to do occasionally for convenience, but prefer to send them on their own in general. 

They're easier to review quickly, and don't "litter" the actual point of the PR. See:
https://drive.google.com/file/d/1CNUwfkvB00ZPeAb7aVssF9mIYINzumeZ/view?usp=sharing

You can just reply ACK and resolve this comment.

File: src/main/java/com/google/sps/servlets/ContestsServlet.java [hunk 3]
```diff
@@ -62,30 +64,30 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
     }
 
     Election election = Election.fromEntity(electionEntityOptional.get());
-    
-    List<String> contestJsonList =
+
+    List<JsonElement> contestJsonList =
```
Comment (julian-yang) [inline src/main/java/com/google/sps/servlets/ContestsServlet.java:68]: Can you put the JsonParser.parseString on a separate map() call?

File: src/main/java/com/google/sps/servlets/ContestsServlet.java [hunk 4]
```diff
@@ -62,30 +64,30 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
         ImmutableSet.copyOf(election.getReferendums())
             .stream()
             .map(id -> KeyFactory.createKey(Referendum.ENTITY_KIND, id.longValue()))
             .map(key -> ServletUtils.getFromDatastore(datastore, key))
-            .map(entity -> Referendum.fromEntity(entity).toJsonString())
+            .map(entity -> JsonParser.parseString(Referendum.fromEntity(entity).toJsonString()))
             .collect(ImmutableList.toImmutableList());
 
     // The toJson function reserializes the already serialized JSON strings, so remove the extra
     // escaped characters.
```
Comment (julian-yang) [inline src/main/java/com/google/sps/servlets/ContestsServlet.java:87]: Remove?

File: src/main/java/com/google/sps/data/Candidate.java [hunk 5]
```diff
@@ -27,6 +27,7 @@
   public static final String ENTITY_KIND = "Candidate";
   public static final String NAME_JSON_KEYWORD = "name";
   public static final String PARTY_JSON_KEYWORD = "party";
+  public static final String CAMPAIGN_JSON_KEYWORD = "candidateUrl";
```
Comment (rogerlaucorp) [inline src/main/java/com/google/sps/data/Candidate.java:30]: Prefer more descriptive variable names, especially in java land. 

Change this to CANDIDATE_CAMPAIGN_URL_JSON_KEYWORD. CAMPAIGN_JSON_KEYWORD makes me do a double take when I see that this is actually about a candidate's URL.

File: src/test/java/com/google/sps/data/CandidateTest.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+        new JSONObject(
+            "{\""
+                + Candidate.NAME_JSON_KEYWORD
+                + "\":\"Jane Doe\",\""
+                + Candidate.PARTY_JSON_KEYWORD
+                + "\":\"Green Party\",\""
+                + Candidate.CAMPAIGN_JSON_KEYWORD
+                + "\":\"www.janedoe.org\"}");
+
+    System.out.println(candidateJSON);
```
Comment (rogerlaucorp) [inline src/test/java/com/google/sps/data/CandidateTest.java:49]: Remove debugging statements like this before sending for review.

File: src/test/java/com/google/sps/data/ContestTest.java [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+    DatastoreService ds = DatastoreServiceFactory.getDatastoreService();
+    Contest newContest = Contest.fromJSONObject(ds, contestJSON);
+    Set<Long> candidateIds = newContest.getCandidates();
+    Assert.assertEquals(candidateIds.size(), 2);
+    Assert.assertEquals(newContest.getName(), "Governer");
+    Assert.assertEquals(newContest.getDescription(), "");
+  }
+
+  @Test
+  public void oneContest_twoCandidates_JsonString() {
```
Comment (julian-yang) [inline src/test/java/com/google/sps/data/ContestTest.java:66]: nit: toJsonString

File: src/test/java/com/google/sps/data/ContestTest.java [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+                + "],\""
+                + Contest.DESCRIPTION_ENTITY_KEYWORD
+                + "\":\"Race for California Governer\"}");
+
+    Assert.assertEquals(expectedContestJSON, returnedContestJSON);
+  }
+
+  @Test
+  public void testFromEntity() {
+    ImmutableSet<Long> idSet = ImmutableSet.<Long>builder().add(1L).add(2L).build();
```
Comment (rogerlaucorp) [inline src/test/java/com/google/sps/data/ContestTest.java:118]: I think you can just do:

ImmutableSet.of(1L, 2L);

for small numbers of elements.

File: src/test/java/com/google/sps/data/ContestTest.java [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+    Contest newContest = Contest.fromEntity(newContestEntity);
+
+    Assert.assertEquals("Governer", newContest.getName());
+    Assert.assertEquals(idSet, newContest.getCandidates());
+    Assert.assertEquals("Race for California Governer", newContest.getDescription());
+  }
+
+  @Test
+  public void testAddToDatastore() throws Exception {
+    ImmutableSet<Long> idSet = ImmutableSet.<Long>builder().add(1L).add(2L).build();
```
Comment (rogerlaucorp) [inline src/test/java/com/google/sps/data/ContestTest.java:134]: Same here.

File: .gitignore (hunk 2 above)
Comment (gianelgado12) [reply .gitignore:15]: ACK

File: src/test/java/com/google/sps/data/ElectionTest.java [hunk 10]
```diff
@@ -150,11 +150,11 @@ public void testAddVoterInfoQueryData() throws Exception {
...
             "{\"election\": {\"id\": \"9999\"},"
                 + "\"contests\": [{\"type\": \"type1\",\"office\": \"officeName\","
-                + "\"candidates\": [{\"name\": \"name1\",\"party\": \"party1\"},"
-                + "{\"name\": \"name2\",\"party\": \"party2\"}]},"
+                + "\"candidates\": [{\"name\": \"name1\",\"party\": \"party1\",\"candidateUrl\": \"www.siteOne.com\"},"
+                + "{\"name\": \"name2\",\"party\": \"party2\",\"candidateUrl\": \"www.siteOne.com\"}]},"
                 + "{\"type\": \"type2\",\"office\": \"officeName\","
-                + "\"candidates\": [{\"name\": \"name1\",\"party\": \"party1\"},"
-                + "{\"name\": \"name2\",\"party\": \"party2\"}]}],"
+                + "\"candidates\": [{\"name\": \"name1\",\"party\": \"party1\",\"candidateUrl\":\"www.siteOne.com\"},"
```
Comment (julian-yang) [inline src/test/java/com/google/sps/data/ElectionTest.java:156]: Does adding these fields affect the Election object generated?

Comment (gianelgado12) [reply src/test/java/com/google/sps/data/ElectionTest.java:156]: The actual JSON returned by voterInfoQuery returns a field for the candidateUrl so what I did is parse that field in Candidate.java and as a result if the field is not present a JSONException is thrown by Candidate.fromJSONObject.

Comment (gianelgado12) [reply src/test/java/com/google/sps/data/ElectionTest.java:156]: Its possible the JSON returned by voterInfoQuery could be missing that field so the plan is to account for that in a follow up PR.

File: src/main/java/com/google/sps/data/Candidate.java (hunk 5 above)
Comment (gianelgado12) [reply src/main/java/com/google/sps/data/Candidate.java]: done

File: src/test/java/com/google/sps/data/CandidateTest.java (hunk 6 above)
Comment (gianelgado12) [reply src/test/java/com/google/sps/data/CandidateTest.java]: done

File: src/test/java/com/google/sps/data/ContestTest.java (hunk 8 above)
Comment (gianelgado12) [reply src/test/java/com/google/sps/data/ContestTest.java]: done

File: src/test/java/com/google/sps/data/ContestTest.java (hunk 9 above)
Comment (gianelgado12) [reply src/test/java/com/google/sps/data/ContestTest.java]: done

File: src/main/java/com/google/sps/servlets/ContestsServlet.java (hunk 3 above)
Comment (gianelgado12) [reply src/main/java/com/google/sps/servlets/ContestsServlet.java:68]: done

File: src/main/java/com/google/sps/servlets/ContestsServlet.java (hunk 4 above)
Comment (gianelgado12) [reply src/main/java/com/google/sps/servlets/ContestsServlet.java]: done

File: src/test/java/com/google/sps/data/ContestTest.java (hunk 7 above)
Comment (gianelgado12) [reply src/test/java/com/google/sps/data/ContestTest.java]: done

File: src/test/java/com/google/sps/data/ElectionTest.java (hunk 10 above)
Comment (julian-yang) [reply src/test/java/com/google/sps/data/ElectionTest.java:156]: Can you add an extra test that ensures an exception is thrown when the field is not present then? (in the follow up PR is fine)

**Pull request merged**